pjlib/src/pj/ssl_sock_apple.m

```diff
@@ -0,0 +1,2013 @@
+    }
+
+    nw_parameters_set_local_endpoint(parameters, local_endpoint);
+    nw_release(local_endpoint);
+    
+    *p_params = parameters;
+    return PJ_SUCCESS;
+}
+
+/* Setup assock's connection state callback and start the connection */
+static pj_status_t network_setup_connection(pj_ssl_sock_t *ssock,
+					    void *connection)
+{
+    applessl_sock_t *assock = (applessl_sock_t *)ssock;
+    assock->connection = (nw_connection_t)connection;
+    pj_status_t status;
+
+    /* Initialize input circular buffer */
+    status = circ_init(ssock->pool->factory, &ssock->circ_buf_input, 8192);
+    if (status != PJ_SUCCESS)
+        return status;
+
+    /* Initialize output circular buffer */
+    status = circ_init(ssock->pool->factory, &ssock->circ_buf_output, 8192);
+    if (status != PJ_SUCCESS)
+        return status;
+
+    /* Hold a reference until cancelled */
+    nw_retain(assock->connection);
+    nw_connection_set_queue(assock->connection, assock->queue);
+
+    nw_connection_set_state_changed_handler(assock->connection,
+    	^(nw_connection_state_t state, nw_error_t error)
+    {
+    	pj_status_t status = PJ_SUCCESS;
+    	pj_bool_t call_cb = PJ_FALSE;
+
+	if (error) {
+    	    errno = nw_error_get_error_code(error);
+            warn("Connection failed %p", assock);
+            status = PJ_EEOF;
+            call_cb = PJ_TRUE;	
+	}
+
+	if (state == nw_connection_state_ready) {
+	    if (ssock->is_server) {
+    		nw_protocol_definition_t tls_def;
+    		nw_protocol_metadata_t prot_meta;
+    		sec_protocol_metadata_t meta;
+    		
+    		tls_def = nw_protocol_copy_tls_definition();
+    		prot_meta = nw_connection_copy_protocol_metadata(connection,
+    								 tls_def);
+		meta = nw_tls_copy_sec_protocol_metadata(prot_meta);
+		assock->cipher =
+	      	    sec_protocol_metadata_get_negotiated_tls_ciphersuite(meta);
+
+		if (ssock->param.require_client_cert &&
+		    !sec_protocol_metadata_access_peer_certificate_chain(
+		    	meta, ^(sec_certificate_t certificate) {} ))
+		{
+		    status = PJ_EEOF;
+		}
+		nw_release(tls_def);
+		nw_release(prot_meta);
+		nw_release(meta);
+	    }
+	    call_cb = PJ_TRUE;
+	} else if (state == nw_connection_state_cancelled) {
+	    /* We release the reference in ssl_destroy() */
+	    // nw_release(assock->connection);
+	    // assock->connection = nil;
+	}
+
+	if (call_cb) {
+	    [assock->lock lock];
+	    assock->event.type = EVENT_HANDSHAKE_COMPLETE;
+	    assock->event.body.handshake_ev.status = status;
+	    dispatch_semaphore_signal(assock->cb_semaphore);
+	    dispatch_semaphore_wait(assock->ev_semaphore, DISPATCH_TIME_FOREVER);
+	    [assock->lock unlock];
+
+	    if (ssock->is_server && status == PJ_SUCCESS) {
+    	        status = network_start_read(ssock, ssock->param.async_cnt,
+    		             (unsigned)ssock->param.read_buffer_size,
+			     ssock->asock_rbuf, 0);
+	    }
+	}
+    });
+
+    nw_connection_start(assock->connection);
+    
+    return PJ_SUCCESS;
```

Comment:
for outgoing connection, perhaps EPENDING is better.
but since it's used for incoming connection as well (see https://github.com/pjsip/pjproject/pull/2482/files#diff-90cd42f10cdd2eca70a29c9b97bb23d5 line 964), in ssl_sock_imp_common.c on_accept_complete(), the connection setup can be considered a success/complete.